matcher/src/lib.rs

```diff
@@ -39,30 +90,56 @@ use crate::matrix::MatrixSlab;
 /// `indices` vector if there isn't enough capacity). However this scratch
-/// memory is fairly large (around 135KB) so creating a matcher is expensive and
-/// should be reused.
+/// memory is fairly large (around 135KB) so creating a matcher is expensive.
+///
+/// All `.._match` functions will not compute the indices  of the matched
+/// characters. These should be used to prefitler to filter and rank all
+/// matches. All `.._indices` functions will also compute the indices of the
+/// matched characters but are slower compared to the `..match` variant. These
+/// should be used when rendering the best N matches. Note that the `indices`
+/// argument is **never cleared**. This allows running multiple different
+/// matches on the same haystack and merging the indices by sorting and
+/// deduplicating the vector.
+///
+/// The `needle` argument for each function must always be normalized by the
+/// caller (unicode normalization and case folding). Otherwise, the matcher
+/// may fail to produce a match. The [`pattern`] modules provides utilities
+/// to preprocess needles and **should usually be preferred over invoking the
+/// matcher directly**.  Additionally it's recommend to perform separate matches
+/// for each word in the needle. Consider the folloling example:
+///
+/// If `foo bar` is used as the needle it matches both `foo test baaar` and
+/// `foo hello-world bar`. However, `foo test baaar` will receive a higher
+/// score than `foo hello-world bar`. `baaar` contains a 2 character gap which
+/// will receive a penalty and therefore the user will likely expect it to rank
+/// lower. However, if `foo bar` is matched as a single query `hello-world` and
+/// `test` are both considered gaps too. As `hello-world` is a much longer gap
+/// then `test` the extra penalty for `baaar` is canceled out. If both words
+/// are matched individually the interspersed words do not receive a penalty and
+/// `foo hello-world bar` ranks higher.
 ///
-/// All `.._match` functions will not compute the indices of the matched chars
-/// and are therefore significantly faster. These should be used to prefitler
-/// and sort all matches. All `.._indices` functions will compute the indices of
-/// the computed chars. These should be used when rendering the best N matches.
-/// Note that the `indices` argument is **never cleared**. This allows running
-/// multiple different matches on the same haystack and merging the indices by
-/// sorting and deduplicating the vector.
+/// In general nucleo is a **substring matching tool** (except for the prefix/
+/// postfix matching modes) with no penalty assigned to matches that start
+/// later within the same pattern (which enables matching words individually
+/// as shown above). If patterns show a large variety in length and the syntax
+/// descibled above is not used it may be preferable to give preference to
+/// matches closer to the start of a haystack.  To accomedate that usecase the
```

Comment:
```suggestion
/// described above is not used it may be preferable to give preference to
/// matches closer to the start of a haystack. To accomodate that usecase the
```